### PR TITLE
Issue 9631 - Error message not using fully qualified name

### PR DESCRIPTION
**Pull request by ntrel**
Fix *cannot implicitly convert expression of type T to T* ambiguous errors.
Add `Type.toAutoQualChars()`.

There are other related errors that could use this fix, this only addresses the first test case in bugzilla.

### CODE REVIEW

File: src/ddmd/mtype.d [hunk 1]
```diff
@@ -910,6 +910,18 @@ extern (C++) abstract class Type : RootObject
         return buf.extractString();
     }
 
+    static const(char*)[2] toAutoQualChars(Type t1, Type t2)
```
Comment (JinShil) [inline src/ddmd/mtype.d]: Please add DDoc comment header.

Comment (JinShil) [reply src/ddmd/mtype.d]: Also, why the word "Auto"?  I would have expected something like `toFullyQualChars`.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/ddmd/mtype.d (hunk 1 above)
Comment (ntrel) [reply src/ddmd/mtype.d]: Will do. Auto is because it only qualifies the types if they would otherwise collide. Suggestions welcome.

Comment (JinShil) [reply src/ddmd/mtype.d]: Ok, I see.  No need to change it in my opinion.  I was just curios.

**Review by WalterBright — CHANGES_REQUESTED** [review]
Needs Ddoc header for toAutoQualChars() function. Also, please make it a non-member function instead of `static`.

File: src/ddmd/mtype.d [hunk 2]
```diff
@@ -9171,3 +9171,19 @@ extern (C++) final class Parameter : RootObject
 
     extern (D) static immutable bool[SR.max + 1][SR.max + 1] covariant = covariantInit();
 }
+
+/***********************************************************
+ * For printing two types with qualification when necessary.
+ * Disambiguates the case where toChars() on each type match.
+ */
+const(char*)[2] toAutoQualChars(Type t1, Type t2)
```
Comment (JinShil) [inline src/ddmd/mtype.d]: Still needs "Returns" and "Params".  See https://dlang.org/spec/ddoc.html

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/ddmd/mtype.d (hunk 2 above)
Comment (ntrel) [reply src/ddmd/mtype.d]: Added `Returns`. I don't know anything meaningful to put for `Params`.

File: src/ddmd/mtype.d [hunk 3]
```diff
@@ -9171,3 +9172,20 @@ extern (C++) final class Parameter : RootObject
 
     extern (D) static immutable bool[SR.max + 1][SR.max + 1] covariant = covariantInit();
 }
+
+/***********************************************************
+ * For printing two types with qualification when necessary.
+ * Disambiguates the case where toChars() on each type match.
+ * Returns: Names of types.
+ */
+const(char*)[2] toAutoQualChars(Type t1, Type t2)
```
Comment (JinShil) [inline src/ddmd/mtype.d]: How about this?
```D
/**
* For printing two types with qualification when necessary.
* Params:
*    t1 = The first type to receive the type name for
*    t2 = The second type to receive the type name for
* Returns:  
*    The fully-qualified names of both types if the two type names are not the same, 
*    or the unqualified names of both types if the two type names are the same.
*/

Comment (ntrel) [reply src/ddmd/mtype.d]: Added.

**Review by JinShil — APPROVED** [review]

**Comment by ntrel** [comment]
BTW I have follow up PRs for 3 other kinds of error message where the same fix applies.

File: test/fail_compilation/bug9631.d [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+template T2()
+{
+    struct F { }
+}
+
+void main()
+{
+    T2!().F x = T1!().F();
```
Comment (JinShil) [inline test/fail_compilation/bug9631.d]: Sorry @ntrel but this isn't passing coverage.  You'll have to add a test that tests the case when both `t1` and `t2` are the same.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: test/fail_compilation/bug9631.d (hunk 4 above)
Comment (JinShil) [reply test/fail_compilation/bug9631.d]: Nevermind, The coverage report looks weird, and I project coverage looks good, so let's merge this.

**Pull request merged**